flink-table-store-core/src/main/java/org/apache/flink/table/store/table/ChangelogWithKeyFileStoreTable.java

```diff
@@ -142,15 +145,40 @@ protected void withNonPartitionFilter(Predicate predicate) {
                                 tableSchema.fieldNames(),
                                 tableSchema.trimmedPrimaryKeys());
                 if (keyFilters.size() > 0) {
-                    scan.withKeyFilter(PredicateBuilder.and(keyFilters));
+                    scan.withKeyFilter(and(keyFilters));
                 }
             }
         };
     }
 
     @Override
     public TableRead newRead() {
+        List<String> primaryKeys = tableSchema.trimmedPrimaryKeys();
+        Set<String> nonPrimaryKeys =
+                tableSchema.fieldNames().stream()
+                        .filter(name -> !primaryKeys.contains(name))
+                        .collect(Collectors.toSet());
         return new KeyValueTableRead(store.newRead()) {
+
+            @Override
+            public TableRead withFilter(Predicate predicate) {
+                List<Predicate> predicates = new ArrayList<>();
+                for (Predicate sub : splitAnd(predicate)) {
+                    // TODO support value filter
+                    if (containsFields(sub, nonPrimaryKeys)) {
+                        continue;
+                    }
+
+                    // TODO Actually, the index is wrong, but it is OK. The orc filter
+                    // just use name instead of index.
+                    predicates.add(sub);
```

Comment:
Are the names really matched? Key field names are prefixed with `_KEY_` but I suspect field names in predicate don't have this prefix.

Reply:
We don't need to max keys.
Actually, we are filtering fields in value. 
One plan is to remove the key fields from the file, since they can all be retrieved from the value. See FLINK-27103